usaspending_api/references/v2/views/filter_tree/psc_filter_tree.py

```diff
@@ -15,18 +15,215 @@
+    },
+    # 0 - 9
+    "Product": {"pattern": r"^\d\d\d\d$", "expanded_terms": [digit for digit in digits]},
+}
+
 
 class PSCFilterTree(FilterTree):
-    def raw_search(self, tiered_keys, filter_string=None):
+    def raw_search(self, tiered_keys, child_layers, filter_string):
         if not self._path_is_valid(tiered_keys):
             return []
-
         if len(tiered_keys) == 0:
-            return self._toptier_search()
+            if child_layers != 0:
+                if child_layers == 2 or child_layers == -1:
+                    tier3_nodes = self.tier_3_search(tiered_keys, filter_string)
+                    tier2_nodes = self.tier_2_search(tiered_keys, filter_string, tier3_nodes)
+                    tier2_nodes = self._combine_nodes(tier2_nodes, tier3_nodes)
+                    tier1_nodes = self.tier_1_search(tiered_keys, filter_string, tier2_nodes)
+                    tier1_nodes = self._combine_nodes(tier1_nodes, tier2_nodes)
+                    toptier_nodes = self.toptier_search(filter_string, tier1_nodes)
+                    toptier_nodes = self._combine_nodes(toptier_nodes, tier1_nodes)
+                else:
+                    tier1_nodes = self.tier_1_search(tiered_keys, filter_string)
+                    toptier_nodes = self.toptier_search(filter_string, tier1_nodes)
+                toptier_nodes = self._combine_nodes(toptier_nodes, tier1_nodes)
+            else:
+                toptier_nodes = self.toptier_search(filter_string)
+            retval = toptier_nodes
         elif len(tiered_keys) == 1:
-            return self._psc_from_group(tiered_keys[0])
+            if child_layers != 0:
+                tier3_nodes = self.tier_3_search(tiered_keys, filter_string)
+                tier2_nodes = self.tier_2_search(tiered_keys, filter_string, tier3_nodes)
+                tier2_nodes = self._combine_nodes(tier2_nodes, tier3_nodes)
+                tier1_nodes = self.tier_1_search(tiered_keys, filter_string, tier2_nodes)
+                tier1_nodes = self._combine_nodes(tier1_nodes, tier2_nodes)
+            else:
+                tier1_nodes = self.tier_1_search(tiered_keys, filter_string)
+            retval = tier1_nodes
+        elif len(tiered_keys) == 2:
+            if child_layers != 0:
+                tier3_nodes = self.tier_3_search(tiered_keys, filter_string)
+                tier2_nodes = self.tier_2_search(tiered_keys, filter_string, tier3_nodes)
+                tier2_nodes = self._combine_nodes(tier2_nodes, tier3_nodes)
+                retval = tier2_nodes
+            else:
+                if tiered_keys[0] == "Product":
+                    retval = self.tier_3_search(tiered_keys, filter_string)
+                else:
+                    retval = self.tier_2_search(tiered_keys, filter_string)
         else:
-            return self._psc_from_parent(tiered_keys[-1], filter_string)
+            retval = self.tier_3_search(tiered_keys, filter_string)
+        return retval
```

Comment:
If I understand correctly, the reason for all these if/else is that the new algorithm gets the lowest level of nodes first, then constructs the layers above, rather than top down like before. This code seems to be exhaustively checking each combination of starting and ending depth,  but I'm pretty sure you could just add the two to figure out your bottom layer, then build the pieces you need. It would cut down the CC of this code by a lot.